### PR TITLE
Generate non-written to benchmarking inputs from existing memory instead of new allocation

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122643
* __->__ #121999
* #124031
* #122825
* #123229
* #122642
* #124030

In MAX_AUTOTUNE, we autotune between triton templates at lowering time. This involves allocating tensors to run the benchmarks. Especially when compiling the backward, this can increase peak memory, as we still are holding all of the activations. `benchmark-fusion` exacerbates this because the epilogue-fused mms have more inputs and potentially more outputs. 

See, this comparison of memory when I enable benchmark-fusion just for TritonTemplates.

image

We can avoid this by taking memory for the inputs that are not written to in our benchmark from parameters, instead of doing fresh allocations. Haven't done a full benchmark run yet but in local testing this fixes the memory regression. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -3880,18 +3881,53 @@ def __del__(self):
...
+
         else:
             template_node = nodes[0]
             epilogue_nodes = nodes[1:]
 
-            with config.patch("benchmark_kernel", True):
-                src_code = self.codegen_template(
-                    template_node, epilogue_nodes, only_gen_src_code=True
+            # TODO - handle epilogue inputs as well
+            non_out_arg_indices = list(
```
Comment (shunting314) [inline torch/_inductor/codegen/triton.py:3896]: This part of code in unnecessary since you overwrite `non_out_arg_indices` below?

**Review by shunting314 — COMMENTED** [review]
So this PR is like we 'steal' storage from weights/activations for benchmarking inputs if the inputs are not mutated:) ?

**Review by shunting314 — APPROVED** [review]

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/codegen/triton.py:3896]: Oh yea, this was from an old version. will delete

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -170,6 +172,99 @@ def load_triton():
...
+
+        if not mod_tensor.device == device:
+            continue
+
+        if not mod_tensor.untyped_storage().nbytes() >= needed_size:
+            continue
+
+        # We dont want to use the same storage for multiple example inputs,
+        # would distort timing.
+        stor_data_ptr = mod_tensor.untyped_storage().data_ptr()
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/utils.py:239]: storage_offset() is the difference

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**